src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -521,6 +514,13 @@ void D4NFilterObject::set_attrs_from_obj_state(const DoutPrefixProvider* dpp, op
 
   bl_val.append(this->get_bucket()->get_name());
   attrs[RGW_CACHE_ATTR_BUCKET_NAME] = std::move(bl_val);
+  
+  if (dirty) {
```

Comment:
I have one generic comment for this PR: since you do not add an xattr for a clean block in handle_data(), lets do the same everywhere. We can avoid a call to cache by not setting an xattr in case of a clean block. So modify set_attrs_from_obj_state() to set the xattr for a head block only when it is dirty. The absence of the xattr signifies that the block is clean. Make bool dirty=false as the default argument to this method, so that you just need to pass in true only for dirty blocks. Also make changes while restoring data, only if the xattr is present and set to 1, then set that block to dirty, else set the block to clean.

Reply:
Also what is the status of testing of this PR? Have you tested these changes? and the data restoration also? Let me know.

Reply:
I have tested the PR with lazy deletion for both unversioned and versioned buckets (suspended and enabled), and I tested data restoration as well as restoration during a delete before the cleaning cycle. The data is handled correctly and the cache entries are deleted as expected during the cleaning cycles.

Reply:
Also please test 'get' requests, after a 'put' for both versioned and unversioned buckets if you haven't done so.

Reply:
You can also ensure that the 'dirty' xattr is set by using `getfattr -d -- <filename>`.